doc/sphinx/proof-engine/ltac.rst

```diff
@@ -451,178 +461,185 @@ structure:
+
    .. tacv:: solve @ltac_expr
 
-      This is an |Ltac| alias that gives a primitive access to the :n:`solve:`
-      tactical. See the :n:`first` tactical for more information.
+      This is an |Ltac| alias that gives a primitive access to the :tacn:`solve`
+      tactic. See the :tacn:`first` tactic for more information.
 
-Identity
-~~~~~~~~
+Print/identity tactic
+~~~~~~~~~~~~~~~~~~~~~
 
-The constant :n:`idtac` is the identity tactic: it leaves any goal unchanged but
-it appears in the proof script.
 
-.. tacn:: idtac {* @message_token}
+.. tacn:: idtac {* {| @ident | @string | @int } }
    :name: idtac
 
-   This prints the given tokens. Strings and integers are printed
+   Leaves the proof unchanged and prints the given tokens. Strings and integers are printed
    literally. If a (term) variable is given, its contents are printed.
 
+   :tacn:`idtac` is an :token:`ltac_expr1`.
+
 Failing
 ~~~~~~~
 
-.. tacn:: fail
-   :name: fail
+.. tacn:: {| fail | gfail } {? @int_or_var } {* {| @ident | @string | @int } }
+   :name: fail; gfail
 
-   This is the always-failing tactic: it does not solve any
-   goal. It is useful for defining other tacticals since it can be caught by
+   :tacn:`fail` is the always-failing tactic: it does not solve any
+   goal. It is useful for defining other tactics since it can be caught by
    :tacn:`try`, :tacn:`repeat`, :tacn:`match goal`, or the branching tacticals.
 
-   .. tacv:: fail @num
+   :tacn:`gfail` fails even when used after :n:`;` and there are no goals left.
+   Similarly, :tacn:`gfail` fails even when used after ``all:`` and there are no
+   goals left.
+
+   :tacn:`fail` and :tacn:`gfail` are :token:`ltac_expr1`\s.
+
+
+   See the example for a comparison of the two constructs.
 
-      The number is the failure level. If no level is specified, it defaults to 0.
+   Note that if Coq terms have to be
+   printed as part of the failure, term construction always forces the
+   tactic into the goals, meaning that if there are no goals when it is
+   evaluated, a tactic call like :tacn:`let` :n:`x := H in` :tacn:`fail` `0 x` will succeed.
+
+   :n:`@int_or_var`
+      The failure level. If no level is specified, it defaults to 0.
       The level is used by :tacn:`try`, :tacn:`repeat`, :tacn:`match goal` and the branching
       tacticals. If 0, it makes :tacn:`match goal` consider the next clause
       (backtracking). If nonzero, the current :tacn:`match goal` block, :tacn:`try`,
       :tacn:`repeat`, or branching command is aborted and the level is decremented. In
       the case of :n:`+`, a nonzero level skips the first backtrack point, even if
-      the call to :n:`fail @num` is not enclosed in a :n:`+` command,
+      the call to :tacn:`fail` :n:`@num` is not enclosed in a :n:`+` construct,
       respecting the algebraic identity.
 
-   .. tacv:: fail {* @message_token}
-
+   :n:`{* {| @ident | @string | @int } }`
       The given tokens are used for printing the failure message.
```

Comment:
Should make specific that `ident` represent Ltac variables. And will raise `@ident not found` if the Ltac variable is undefined. Maybe worth improving the description of this `ident` in the `idtac` description too.